projects/packages/connection/tests/php/test_jetpack_xmlrpc_server.php

```diff
@@ -0,0 +1,438 @@
+		Jetpack_Options::update_options(
+			array(
+				'id'         => 1001,
+				'blog_token' => '123456.123456',
+			)
+		);
+
+		$server   = $this->get_mocked_xmlrpc_server();
+		$response = $server->remote_connect(
+			array(
+				'nonce'      => '1234',
+				'local_user' => $this->xmlrpc_admin,
+			),
+			$this->get_mocked_ixr_client( true, false )
+		);
+
+		$this->assertInstanceOf( 'IXR_Error', $response );
+		$this->assertObjectHasAttribute( 'code', $response );
+		$this->assertObjectHasAttribute( 'message', $response );
+		$this->assertEquals( 400, $response->code );
+		$this->assertEquals(
+			'Jetpack: [token_fetch_failed] Failed to fetch user token from WordPress.com.',
+			$response->message
+		);
+	}
+
+	/**
+	 * Test test_xmlrpc_get_user_by_id
+	 */
+	public function test_remote_connect_success() {
+		Jetpack_Options::update_options(
+			array(
+				'id'         => 1001,
+				'blog_token' => '123456.123456',
+			)
+		);
+
+		$server   = $this->get_mocked_xmlrpc_server();
+		$response = $server->remote_connect(
+			array(
+				'nonce'      => '1234',
+				'local_user' => $this->xmlrpc_admin,
+			),
+			$this->get_mocked_ixr_client( true, 'this_is.a_token' )
+		);
+
+		$this->assertTrue( $response );
+	}
+
+	/*
+	 * Helpers
+	 */
+
+	/**
+	 * Return an "ok" status.
+	 *
+	 * @return array
+	 */
+	public function return_ok_status() {
+		return array(
+			'body'     => 'OK',
+			'response' => array(
+				'code'    => 200,
+				'message' => '',
+			),
+		);
+	}
+
+	/**
+	 * Return an "invalid nonce" status.
+	 *
+	 * @return array
+	 */
+	public function return_invalid_nonce_status() {
+		return array(
+			'body'     => 'FAIL: NOT OK',
+			'response' => array(
+				'code'    => 200,
+				'message' => '',
+			),
+		);
+	}
+
+	/**
+	 * Return an "nonce 404" status.
+	 *
+	 * @return array
+	 */
+	public function return_nonce_404_status() {
+		return array(
+			'body'     => '',
+			'response' => array(
+				'code'    => 404,
+				'message' => '',
+			),
+		);
+	}
+
+	/**
+	 * Get a mocked IXR client.
+	 *
+	 * @param bool   $query_called Whether `query` should be called.
+	 * @param string $response Return value for `getResponse`.
+	 * @param bool   $query_return Return value for `query`.
+	 * @param string $error Return value for `isError`.
+	 * @return Jetpack_IXR_Client
+	 */
+	protected function get_mocked_ixr_client( $query_called = false, $response = '', $query_return = true, $error = null ) {
+		$xml = $this->getMockBuilder( 'Jetpack_IXR_Client' )
+			->setMethods(
+				array(
+					'query',
+					'isError',
+					'getResponse',
+				)
+			)
+			->getMock();
+
+		$xml->expects( $this->exactly( $query_called ? 1 : 0 ) )
+			->method( 'query' )
+			->will( $this->returnValue( $query_return ) );
+
+		$xml->expects( $this->exactly( $query_called ? 1 : 0 ) )
+			->method( 'isError' )
+			->will( $this->returnValue( empty( $error ) ? false : true ) );
+
+		$xml->expects( $this->exactly( empty( $error ) ? 1 : 0 ) )
```

Comment:
You can feel free to ignore, but unnecessary use of `empty()` is one of my pet peeves. p1614091140044600-slack-C6JKZ8MLK
```suggestion
		$xml->expects( $this->exactly( $error ? 0 : 1 ) )
```